constantine/trusted_setups/ethereum_kzg_srs.nim

```diff
@@ -221,29 +221,29 @@ proc load_ckzg4844(ctx: ptr EthereumKZGContext, f: File): TrustedSetupStatus =
 
   block:
     # G1 points - 96 characters + newline
-    var bufG1Hex {.noInit.}: array[2*g1Bytes, char]
+    var bufG1Hex {.noInit.}: array[2*g1Bytes+1, char] # On MacOS, an extra byte seems to be needed for fscanf or AddressSanitizer complains
```

Comment:
This is a good find by the sanitizer. It's right. Because you're reading a string (rather than a byte like it's done in c-kzg-4844) it will add a null terminator at the end.

Reply:
I actually open files in binary mode not text: https://github.com/mratsim/constantine/blob/c73ce0e6a0662e905ffc3a670f222818aebe3bfe/constantine/platforms/fileio.nim#L40-L57

And the modifier for strings is %s, which doesn't appear in `const readHexG1 = "%96[^\r\n]%n%*[\r\n]"`
In fact, the sanitizer on Linux doesn't complain.

Reply:
It shouldn't matter that it's opened in binary mode. It just matters that it's reading a string.

Here's an example with `sscanf` which should operate the same:

```c
#include <stdio.h>

int main(void)
{
    char buffer[97];
    int i, n;
    const char *staticString = \
        "0123456789" // 10
        "0123456789" // 20
        "0123456789" // 30
        "0123456789" // 40
        "0123456789" // 50
        "0123456789" // 60
        "0123456789" // 70
        "0123456789" // 80
        "0123456789" // 90
        "012345";    // 96

    for (i = 0; i < 97; i++) {
        buffer[i] = 0xFF;
    }

    printf("\nInitial buffer contents:");
    for (i = 0; i < 97; i++) {
        if (i % 10 == 0) printf("\n");
        printf("%02X ", (unsigned char)buffer[i]);
    }
    printf("\n");

    sscanf(staticString, "%96[^\r\n]%n%*[\r\n]", buffer, &n);

    printf("\nBuffer contents after sscanf:");
    for (i = 0; i < 97; i++) {
        if (i % 10 == 0) printf("\n");
        printf("%02X ", (unsigned char)buffer[i]);
    }
    printf("\n");

    printf("\nNumber of characters read: %d\n", n);

    return 0;
}
```

For me this is the output:

```
$ make test && ./test
cc  -I/opt/homebrew/include/ -L/opt/homebrew/include/ -L/opt/homebrew/lib/  test.c   -o test

Initial buffer contents:
FF FF FF FF FF FF FF FF FF FF
FF FF FF FF FF FF FF FF FF FF
FF FF FF FF FF FF FF FF FF FF
FF FF FF FF FF FF FF FF FF FF
FF FF FF FF FF FF FF FF FF FF
FF FF FF FF FF FF FF FF FF FF
FF FF FF FF FF FF FF FF FF FF
FF FF FF FF FF FF FF FF FF FF
FF FF FF FF FF FF FF FF FF FF
FF FF FF FF FF FF FF

Buffer contents after sscanf:
30 31 32 33 34 35 36 37 38 39
30 31 32 33 34 35 36 37 38 39
30 31 32 33 34 35 36 37 38 39
30 31 32 33 34 35 36 37 38 39
30 31 32 33 34 35 36 37 38 39
30 31 32 33 34 35 36 37 38 39
30 31 32 33 34 35 36 37 38 39
30 31 32 33 34 35 36 37 38 39
30 31 32 33 34 35 36 37 38 39
30 31 32 33 34 35 00

Number of characters read: 96
```

As you can see, the 97th byte has been set to 0 but only 96 bytes were read.

Reply:
> And the modifier for strings is `%s`, which doesn't appear in `const readHexG1 = "%96[^\r\n]%n%*[\r\n]"`
In fact, the sanitizer on Linux doesn't complain.

Oops didn't respond to this. It doesn't need to be `%s` for it to be recognized as a string.